src/transformStateWithClones.js

```
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const arrayClones = [];
+
+  let objectClone = Object.assign({}, state);
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      const cloneAddingProperties = Object.assign(
+        {},
+        objectClone,
+        action.extraData,
+      );
+
+      arrayClones.push(cloneAddingProperties);
+      objectClone = cloneAddingProperties;
+    } else if (action.type === 'removeProperties') {
+      const cloneRemovedProperties = Object.assign({}, objectClone);
+
+      for (const item of action.keysToRemove) {
+        delete cloneRemovedProperties[item];
+      }
+
+      arrayClones.push(cloneRemovedProperties);
+      objectClone = cloneRemovedProperties;
+    } else if (action.type === 'clear') {
+      arrayClones.push({});
+      objectClone = {};
+    }
```

There is no default/error handling for unknown action types. The checklist requires: "3. [CODE STYLE]: switch/case should always have default case for error handling." Add a `default` case (for example, throw `new Error('Unknown action type: ' + action.type)`) so unknown actions are not silently ignored and to ensure the returned array length matches `actions.length` .